### PR TITLE
Solaris 10 Fixes

### LINKED ISSUE #1193: psutil only returns uids of user-owned process on Solaris 10
How to reproduce:

As normal user, chose a PID of another user:

    >>> p = psutil.Process(12345)
    >>> p.uids()
    Traceback (most recent call last):
      File "/path/to/venv/lib/python3.5/site-packages/psutil/_pssunos.py", line 346, in wrapper
        return fun(self, *args, **kwargs)
      File "/path/to/venv/lib/python3.5/site-packages/psutil/_pssunos.py", line 477, in uids
        real, effective, saved, _, _, _ = self._proc_cred()
      File "/path/to/venv/lib/python3.5/site-packages/psutil/_common.py", line 336, in wrapper
        return fun(self)
      File "/path/to/venv/lib/python3.5/site-packages/psutil/_pssunos.py", line 400, in _proc_cred
        return cext.proc_cred(self.pid, self._procfs_path)
    PermissionError: [Errno 13] Permission denied: '/proc/12345/cred'
     
    During handling of the above exception, another exception occurred:
     
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/path/to/venv/lib/python3.5/site-packages/psutil/_common.py", line 336, in wrapper
        return fun(self)
      File "/path/to/venv/lib/python3.5/site-packages/psutil/__init__.py", line 768, in uids
        return self._proc.uids()
      File "/path/to/venv/lib/python3.5/site-packages/psutil/_pssunos.py", line 362, in wrapper
        raise AccessDenied(self.pid, self._name)
    psutil.AccessDenied: psutil.AccessDenied (pid=12345)

The thing is that psutil tries to read the uids from `/proc/$pid/cred` that is always only user-readable on Solaris 10:

    $ ls /proc/12345/cred -l
    -r-------- 1 juser juser 52 2017-11-23 11:27 /proc/12345/cred

A better choice on Solaris 10 is to use `/proc/$pid/psinfo` since it [also contains the uids](https://docs.oracle.com/cd/E19253-01/816-5174/proc-4/index.html) (modulo the saved user id) - and is readable by everyone. Thus, if psutil really cares about the saved user id, it could also try `/proc/$pid/cred` first and then fall back to `/proc/$pid/psinfo`.

For our example:

    $ ls /proc/12345/psinfo -l
    -r--r--r-- 1 juser juser 336 2017-11-23 11:27 /proc/12345/psinfo

### PR DESCRIPTION
**Pull request by gsauthof**
Fixes #1193 and #1194, a double free, some undefined behavior and some warnings.

### CODE REVIEW

File: psutil/_pssunos.py [hunk 1]
```diff
@@ -432,27 +439,10 @@ def num_threads(self):
...
-            # 48 is 'operation not supported' but errno does not expose
-            # it. It occurs for low system pids.
-            if err.errno in (errno.ENOENT, errno.ESRCH, 48):
-                if pid_exists(self.pid):
-                    raise AccessDenied(self.pid, self._name)
-            raise
+        # Note #1: getpriority(3) doesn't work for realtime processes.
+        # Psinfo is what ps uses, see:
+        # https://github.com/giampaolo/psutil/issues/1194
+        return self._proc_basic_info()[proc_info_map['nice']]
```
Comment (giampaolo) [inline psutil/_pssunos.py]: Nice that we got rid of that mess.

File: psutil/_psutil_common.c [hunk 2]
```diff
@@ -40,7 +40,7 @@ PyUnicode_DecodeFSDefaultAndSize(char *s, Py_ssize_t size) {
  * If msg != "" the exception message will change in accordance.
  */
 PyObject *
-NoSuchProcess(char *msg) {
+NoSuchProcess(const char *msg) {
```
Comment (giampaolo) [inline psutil/_psutil_common.c]: why is this necessary?

File: psutil/_psutil_common.h [hunk 3]
```diff
@@ -4,6 +4,9 @@
  * found in the LICENSE file.
  */
 
+#ifndef PSUTIL_PSUTIL_COMMON_H
+#define PSUTIL_PSUTIL_COMMON_H
```
Comment (giampaolo) [inline psutil/_psutil_common.h]: why is this necessary?

File: psutil/_pssunos.py [hunk 4]
```diff
@@ -394,7 +398,10 @@ def _proc_basic_info(self):
 
     @memoize_when_activated
     def _proc_cred(self):
-        return cext.proc_cred(self.pid, self._procfs_path)
+        @wrap_exceptions
```
Comment (giampaolo) [inline psutil/_pssunos.py]: Why not decorate the method instead?

File: psutil/_pssunos.py [hunk 5]
```diff
@@ -471,12 +461,22 @@ def ppid(self):
...
 
     @wrap_exceptions
     def gids(self):
-        _, _, _, real, effective, saved = self._proc_cred()
+        try:
+            _, _, _, real, effective, saved = self._proc_cred()
+        except AccessDenied:
+            real      = self._proc_basic_info()[proc_info_map['gid']]
+            effective = self._proc_basic_info()[proc_info_map['egid']]
+            saved     = None
```
Comment (giampaolo) [inline psutil/_pssunos.py]: PEP8 (please remove extra spaces)

File: psutil/_pssunos.py [hunk 6]
```diff
@@ -471,12 +461,22 @@ def ppid(self):
 
     @wrap_exceptions
     def uids(self):
-        real, effective, saved, _, _, _ = self._proc_cred()
+        try:
+            real, effective, saved, _, _, _ = self._proc_cred()
+        except AccessDenied:
+            real      = self._proc_basic_info()[proc_info_map['uid']]
+            effective = self._proc_basic_info()[proc_info_map['euid']]
+            saved     = None
```
Comment (giampaolo) [inline psutil/_pssunos.py]: PEP8 (please remove extra spaces)

**Review by giampaolo — CHANGES_REQUESTED** [review]
Nice work. Could you please update HISTORY file describing the fixes/improvements which occurred in this PR?

File: psutil/_psutil_common.c (hunk 2 above)
Comment (gsauthof) [reply psutil/_psutil_common.c]: `const char *` means pointer to const characters. The const modifier signals the caller that NoSuchProcess will not change the passed message. Otherwise, it's not clear and one has to inspect the function body. In case someone accidentally writes to msg using that const pointer it's a compile error. This change also remove compiler warnings since NoSuchProcess is called with string literals that effectively have the type `const char *`. Thus, this additional type safety protects against accidental writing into a const character array through a non-const pointer which would invoke undefined behavior (e.g. crash if the string literals are placed in a read-only data section).

File: psutil/_psutil_common.h (hunk 3 above)
Comment (gsauthof) [reply psutil/_psutil_common.h]: This guards the header from being included multiple times into one translation unit (either directly/indirectly). In C, unless the header does something special, you always define such guards (or use the non-standardized once pragma).

File: psutil/_pssunos.py (hunk 4 above)
Comment (gsauthof) [reply psutil/_pssunos.py]: I tried stacking the `@wrap_exceptions` on top of `@memoize_when_activated` of `_proc_cred` but it didn't work. Something about a missing attribute, IIRC.

File: psutil/_pssunos.py (hunk 5 above)
Comment (gsauthof) [reply psutil/_pssunos.py]: Ok, will update (rebase) the commit.

File: psutil/_pssunos.py (hunk 6 above)
Comment (gsauthof) [reply psutil/_pssunos.py]: Ok, will fix this.

**Comment by gsauthof** [comment]
So I've updated the pull request and implemented the requested changes (HISTORY and PEP 8).

Btw, looking at the `HISTORY.rst` file I noticed:

    - 1120_: [SunOS] cmdline() could be truncated at the 15th character when
      reading it from /proc. An extra effort is made by reading it from process
      address space first.

But #1120 doesn't really match the description. Probably a mixup with #694, right?

**Comment by giampaolo** [comment]
> But #1120 doesn't really match the description. Probably a mixup with #694, right?

correct; thanks for the PR.
